features/network_failures.feature

```diff
@@ -0,0 +1,109 @@
+@uses.config.contract_token
+Feature: Ensure network errors are handled gracefully across various services
+
+  Scenario Outline: Various HTTP errors are handled gracefully on attaching contract token
+    # This test simulates various HTTP errors by mocking the response from the serviceclient
+    # when trying to attach contract token
+    Given a `<release>` `<machine_type>` machine with ubuntu-advantage-tools installed
+    # 400 Bad Request
+    When I create a response overlay for `/v1/context/machines/token` with response code `400` and error message `Bad Request`
+    And I append the following on uaclient config:
+      """
+      features:
+        serviceclient_url_responses: "/tmp/response-overlay.json"
+      """
+    When I attempt to attach `contract_token` with sudo
+    Then stderr contains substring:
+      """
+      Error connecting to /v1/context/machines/token: 400 {"error": "Bad Request"}
+      """
+    Then the machine is unattached
+    # 404 Not Found
+    When I create a response overlay for `/v1/context/machines/token` with response code `404` and error message `Not Found`
+    And I append the following on uaclient config:
+      """
+      features:
+        serviceclient_url_responses: "/tmp/response-overlay.json"
+      """
+    When I attempt to attach `contract_token` with sudo
+    Then stderr contains substring:
+      """
+      Error connecting to /v1/context/machines/token: 404 {"error": "Not Found"}
+      """
+    Then the machine is unattached
+    # 503 Bad Gateway
+    When I create a response overlay for `/v1/context/machines/token` with response code `503` and error message `Bad Gateway`
+    And I append the following on uaclient config:
+      """
+      features:
+        serviceclient_url_responses: "/tmp/response-overlay.json"
+      """
+    When I attempt to attach `contract_token` with sudo
+    Then stderr contains substring:
+      """
+      Error connecting to /v1/context/machines/token: 503 {"error": "Bad Gateway"}
+      """
+    Then the machine is unattached
+
+    Examples: ubuntu release
+      | release | machine_type  |
+      | xenial  | lxd-container |
+      | noble   | lxd-container |
+
+  Scenario Outline: Network errors for attaching contract token are handled gracefully
+    # This test simulates network failure by disabling internet connection
+    # and then trying to attach contract token
+    Given a `<release>` `<machine_type>` machine with ubuntu-advantage-tools installed
+    When I disable any internet connection on the machine
+    And I attempt to attach `contract_token` with sudo
+    Then stderr contains substring:
+      """
+      Failed to attach machine. See https://ubuntu.com/pro/dashboard
```

Comment:
I think this actually needs to be fixed for us to consider the respective LP bug closed. The problem here is that you don't have internet connection:
- there is no indication that lack of connection was the problem
- how are people supposed to access their dashboard if the network is down?

Reply:
@a-dubs if you want, we can land the tests and then you change this message on a next step?

Reply:
Oh shoot yeah you're right. That's not exactly a very useful or specific message. So you're right we definitely need to update the message to consider those bugs closed. As such, I think it would be best if I put this back in draft state and make the changes to the source code such that the source code changes and tests are bundled together into one logical commit, so that when I say "fixes lp #XXXX", I mean **_fixes_**